File: frontend/src/lib/api/tracklist.ts
import { api } from "./client";

export type TracklistEntryStatus = "raw" | "resolved" | "acoustid_confirmed";

export interface TracklistEntry {
  id: string;
  position: number;
  start_seconds: number;
  end_seconds: number | null;
  raw_label: string;
  edit_notes: string | null;
  status: TracklistEntryStatus;
  confidence: number | null;
  track_id: string | null;
  mashup_with: string[];
}

export interface ParsedEntry {
  start_seconds: number;
  raw_label: string;
}

export interface ImportJob {
  id: string;
  kind: "paste" | "url_1001tl" | "ocr";
  status: "queued" | "running" | "succeeded" | "failed";
  parsed: ParsedEntry[];
  error: string | null;
  created_at: string;
}

const base = (slug: string) => `/api/sets/${encodeURIComponent(slug)}/tracklist`;

export async function listTracklist(slug: string): Promise<TracklistEntry[]> {
  const { entries } = await api<{ entries: TracklistEntry[] }>(base(slug));
  return entries;
}

export const createEntry = (
  slug: string,
  body: { start_seconds: number; raw_label: string; position?: number },
) =>
  api<TracklistEntry>(`${base(slug)}/entries`, {
    method: "POST",
    body: JSON.stringify(body),
  });

export const patchEntry = (
  slug: string,
  id: string,
  body: Partial<
    Pick<TracklistEntry, "start_seconds" | "end_seconds" | "raw_label" | "edit_notes">
  >,
) =>
  api<TracklistEntry>(`${base(slug)}/entries/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });

export const moveEntry = (slug: string, id: string, afterPosition: number) =>
  api<void>(`${base(slug)}/entries/${id}/move`, {
    method: "PATCH",
    body: JSON.stringify({ after_position: afterPosition }),
  });

export const deleteEntry = (slug: string, id: string) =>
  api<void>(`${base(slug)}/entries/${id}`, { method: "DELETE" });

export const timeShift = (slug: string, afterSeconds: number, deltaSeconds: number) =>
  api<{ affected_count: number }>(`${base(slug)}/time-shift`, {
    method: "POST",
    body: JSON.stringify({ after_seconds: afterSeconds, delta_seconds: deltaSeconds }),
  });

export const importTracklist = (
  slug: string,
  kind: "paste" | "url_1001tl" | "ocr",
  payload: Record<string, unknown>,
) =>
  api<ImportJob>(`${base(slug)}/import`, {
    method: "POST",
    body: JSON.stringify({ kind, payload }),
  });

export const acceptImport = (slug: string, jobId: string, acceptedIndexes: number[]) =>
  api<{ entries: TracklistEntry[] }>(`${base(slug)}/import/${jobId}/accept`, {
    method: "POST",
    body: JSON.stringify({ accepted_indexes: acceptedIndexes }),
  });

File: packages/core/src/setvault_core/schemas/tracklist.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

TracklistEntryStatus = Literal["raw", "resolved", "acoustid_confirmed"]


class TracklistEntryOut(BaseModel):
    id: str
    position: int
    start_seconds: int
    end_seconds: int | None = None
    raw_label: str
    edit_notes: str | None = None
    status: TracklistEntryStatus
    confidence: float | None = None
    track_id: str | None = None
    mashup_with: list[str] = Field(default_factory=list)


class TracklistEntryCreateIn(BaseModel):
    start_seconds: int = Field(ge=0)
    raw_label: str = Field(min_length=1, max_length=1024)
    position: int | None = None
    end_seconds: int | None = None
    edit_notes: str | None = None


class TracklistEntryPatchIn(BaseModel):
    start_seconds: int | None = Field(default=None, ge=0)
    end_seconds: int | None = None
    raw_label: str | None = Field(default=None, min_length=1, max_length=1024)
    edit_notes: str | None = None
    mashup_with: list[str] | None = None


class TracklistEntryMoveIn(BaseModel):
    after_position: int = Field(ge=-1)


class TimeShiftIn(BaseModel):
    after_seconds: int = Field(ge=0)
    delta_seconds: int  # may be negative


class TimeShiftPreviewOut(BaseModel):
    affected_count: int
    new_positions: list[dict]  # [{id, new_start_seconds}]


class ParsedEntry(BaseModel):
    start_seconds: int | None = None
    raw_label: str


class TracklistImportIn(BaseModel):
    kind: Literal["paste", "url_1001tl", "ocr"]
    payload: dict


class TracklistImportOut(BaseModel):
    id: str
    kind: Literal["paste", "url_1001tl", "ocr"]
    status: Literal["queued", "running", "succeeded", "failed"]
    parsed: list[ParsedEntry] = Field(default_factory=list)
    error: str | None = None
    created_at: datetime


class TracklistImportAcceptIn(BaseModel):
    accepted_indexes: list[int]
